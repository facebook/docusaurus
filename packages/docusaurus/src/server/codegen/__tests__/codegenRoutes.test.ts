/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {
  generateRoutesCode,
  genChunkName,
  generateRoutePropFilename,
} from '../codegenRoutes';
import type {RouteConfig} from '@docusaurus/types';

describe('generateRoutePropFilename', () => {
  it('generate filename based on route path', () => {
    expect(
      generateRoutePropFilename({
        path: '/some/route-path/',
        component: '@theme/Home',
      }),
    ).toEqual(expect.stringMatching(/^some-route-path-[a-z\d]{3}.json$/));
  });

  it('generate filename for /', () => {
    expect(
      generateRoutePropFilename({
        path: '/',
        component: '@theme/Home',
      }),
    ).toEqual(expect.stringMatching(/^index-[a-z\d]{3}.json$/));
  });

  it('generate filename for /category/', () => {
    expect(
      generateRoutePropFilename({
        path: '/category/',
        component: '@theme/Home',
      }),
    ).toEqual(expect.stringMatching(/^category-[a-z\d]{3}.json$/));
  });

  it('generate unique filenames for /', () => {
    expect(
      generateRoutePropFilename({path: '/', component: '@theme/Home'}),
    ).toEqual(generateRoutePropFilename({path: '/', component: '@theme/Home'}));
    expect(
      generateRoutePropFilename({path: '/', component: '@theme/Home'}),
    ).not.toEqual(
      generateRoutePropFilename({
        path: '/',
        component: '@theme/AnotherComponent',
      }),
    );
  });

  it('generate unique filenames for /some/path', () => {
    expect(
      generateRoutePropFilename({path: '/some/path', component: '@theme/Home'}),
    ).toEqual(
      generateRoutePropFilename({path: '/some/path', component: '@theme/Home'}),
    );
    expect(
      generateRoutePropFilename({
        path: '/some/path',
        component: '@theme/Home',
      }),
    ).not.toEqual(
      generateRoutePropFilename({
        path: '/some/path',
        component: '@theme/AnotherComponent',
      }),
    );
  });
});

describe('genChunkName', () => {
  it('works', () => {
    const firstAssert: {[key: string]: string} = {
      '/docs/adding-blog': 'docs-adding-blog-062',
      '/docs/versioning': 'docs-versioning-8a8',
      '/': 'index',
      '/blog/2018/04/30/How-I-Converted-Profilo-To-Docusaurus':
        'blog-2018-04-30-how-i-converted-profilo-to-docusaurus-4f2',
      '/youtube': 'youtube-429',
      '/users/en/': 'users-en-f7a',
      '/blog': 'blog-c06',
    };
    Object.keys(firstAssert).forEach((str) => {
      expect(genChunkName(str)).toBe(firstAssert[str]);
    });
  });

  it("doesn't allow different chunk name for same path", () => {
    expect(genChunkName('path/is/similar', 'oldPrefix')).toEqual(
      genChunkName('path/is/similar', 'newPrefix'),
    );
  });

  it('emits different chunk names for different paths even with same preferred name', () => {
    const secondAssert: {[key: string]: string} = {
      '/blog/1': 'blog-85-f-089',
      '/blog/2': 'blog-353-489',
    };
    Object.keys(secondAssert).forEach((str) => {
      expect(genChunkName(str, undefined, 'blog')).toBe(secondAssert[str]);
    });
  });

  it('only generates short unique IDs', () => {
    const thirdAssert: {[key: string]: string} = {
      a: '0cc175b9',
      b: '92eb5ffe',
      c: '4a8a08f0',
      d: '8277e091',
    };
    Object.keys(thirdAssert).forEach((str) => {
      expect(genChunkName(str, undefined, undefined, true)).toBe(
        thirdAssert[str],
      );
    });
    expect(genChunkName('d', undefined, undefined, true)).toBe('8277e091');
  });

  // https://github.com/facebook/docusaurus/issues/8536
  it('avoids hash collisions', () => {
    expect(
      genChunkName(
        '@site/blog/2022-11-18-bye-medium/index.mdx?truncated=true',
        'content',
        'blog',
        false,
      ),
    ).not.toBe(
      genChunkName(
        '@site/blog/2019-10-05-react-nfc/index.mdx?truncated=true',
        'content',
        'blog',
        false,
      ),
    );
    expect(
      genChunkName(
        '@site/blog/2022-11-18-bye-medium/index.mdx?truncated=true',
        'content',
        'blog',
        true,
      ),
    ).not.toBe(
      genChunkName(
        '@site/blog/2019-10-05-react-nfc/index.mdx?truncated=true',
        'content',
        'blog',
        true,
      ),
    );
  });
});

describe('loadRoutes', () => {
  it('loads nested route config', () => {
    const nestedRouteConfig: RouteConfig = {
      component: '@theme/DocRoot',
      path: '/docs:route',
      modules: {
        docsMetadata: 'docs-b5f.json',
      },
      routes: [
        {
          path: '/docs/hello',
          component: '@theme/DocItem',
          exact: true,
          modules: {
            content: 'docs/hello.md',
            metadata: 'docs-hello-da2.json',
          },
          metadata: {
            sourceFilePath: 'docs/hello.md',
            lastUpdatedAt: 1710842708527,
          },
          context: {
            plugin: 'pluginRouteContextModule-100.json',
          },
          sidebar: 'main',
        },
        {
          path: 'docs/foo/baz',
          component: '@theme/DocItem',
          modules: {
            content: 'docs/foo/baz.md',
            metadata: 'docs-foo-baz-dd9.json',
          },
          metadata: {
            sourceFilePath: 'docs/foo/baz.md',
            lastUpdatedAt: 1710842708527,
          },
          context: {
            plugin: 'pluginRouteContextModule-100.json',
          },
          sidebar: 'secondary',
          'key:a': 'containing colon',
          "key'b": 'containing quote',
          'key"c': 'containing double quote',
          'key,d': 'containing comma',
          字段: 'containing unicode',
        },
      ],
    };
    expect(
      generateRoutesCode([nestedRouteConfig], '/', 'ignore'),
    ).toMatchSnapshot();
  });

  it('loads flat route config', () => {
    const flatRouteConfig: RouteConfig = {
      path: '/blog',
      component: '@theme/BlogListPage',
      exact: true,
      metadata: {
        lastUpdatedAt: 1710842708527,
      },
      modules: {
        items: [
          {
            content: {
              __import: true,
              path: 'blog/2018-12-14-Happy-First-Birthday-Slash.md',
              query: {
                truncated: true,
              },
            },
            metadata: 'blog-2018-12-14-happy-first-birthday-slash-d2c.json',
          },
          {
            content: 'blog/2018-12-14-Happy-First-Birthday-Slash.md',
          },
          {
            content: {
              __import: true,
              path: 'blog/2018-12-14-Happy-First-Birthday-Slash.md',
            },
          },
        ],
      },
    };
    expect(
      generateRoutesCode([flatRouteConfig], '/', 'ignore'),
    ).toMatchSnapshot();
  });

  it('rejects invalid route config', () => {
    const routeConfigWithoutPath = {
      component: 'hello/world.js',
    } as RouteConfig;

    expect(() => generateRoutesCode([routeConfigWithoutPath], '/', 'ignore'))
      .toThrowErrorMatchingInlineSnapshot(`
      "Invalid route config: path must be a string and component is required.
      {"component":"hello/world.js"}"
    `);

    const routeConfigWithoutComponent = {
      path: '/hello/world',
    } as RouteConfig;

    expect(() =>
      generateRoutesCode([routeConfigWithoutComponent], '/', 'ignore'),
    ).toThrowErrorMatchingInlineSnapshot(`
      "Invalid route config: path must be a string and component is required.
      {"path":"/hello/world"}"
    `);
  });

  it('loads route config with empty (but valid) path string', () => {
    const routeConfig = {
      path: '',
      component: 'hello/world.js',
    } as RouteConfig;

    expect(generateRoutesCode([routeConfig], '/', 'ignore')).toMatchSnapshot();
  });
});
