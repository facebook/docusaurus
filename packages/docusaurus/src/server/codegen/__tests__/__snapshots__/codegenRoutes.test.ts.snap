// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadRoutes loads flat route config 1`] = `
{
  "registry": {
    "__comp---theme-blog-list-pagea-6-a-7ba": "@theme/BlogListPage",
    "content---blog-0-b-4-09e": "blog/2018-12-14-Happy-First-Birthday-Slash.md?truncated=true",
    "content---blog-7-b-8-fd9": "blog/2018-12-14-Happy-First-Birthday-Slash.md",
    "metadata---blog-0-b-6-74c": "blog-2018-12-14-happy-first-birthday-slash-d2c.json",
  },
  "routesChunkNames": {
    "/blog-030": {
      "__comp": "__comp---theme-blog-list-pagea-6-a-7ba",
      "items": [
        {
          "content": "content---blog-0-b-4-09e",
          "metadata": "metadata---blog-0-b-6-74c",
        },
        {
          "content": "content---blog-7-b-8-fd9",
        },
        {
          "content": "content---blog-7-b-8-fd9",
        },
      ],
    },
  },
  "routesConfig": "import React from 'react';
import ComponentCreator from '@docusaurus/ComponentCreator';

export default [
  {
    path: '/blog',
    component: ComponentCreator('/blog', '030'),
    exact: true
  },
  {
    path: '*',
    component: ComponentCreator('*'),
  },
];
",
}
`;

exports[`loadRoutes loads nested route config 1`] = `
{
  "registry": {
    "__comp---theme-doc-item-178-a40": "@theme/DocItem",
    "__comp---theme-doc-roota-94-67a": "@theme/DocRoot",
    "content---docs-foo-baz-8-ce-61e": "docs/foo/baz.md",
    "content---docs-helloaff-811": "docs/hello.md",
    "docsMetadata---docs-routef-34-881": "docs-b5f.json",
    "metadata---docs-foo-baz-2-cf-fa7": "docs-foo-baz-dd9.json",
    "metadata---docs-hello-956-741": "docs-hello-da2.json",
    "plugin---docs-hello-665-3ca": "pluginRouteContextModule-100.json",
  },
  "routesChunkNames": {
    "/docs/hello-ab4": {
      "__comp": "__comp---theme-doc-item-178-a40",
      "__context": {
        "plugin": "plugin---docs-hello-665-3ca",
      },
      "content": "content---docs-helloaff-811",
      "metadata": "metadata---docs-hello-956-741",
    },
    "/docs:route-001": {
      "__comp": "__comp---theme-doc-roota-94-67a",
      "docsMetadata": "docsMetadata---docs-routef-34-881",
    },
    "docs/foo/baz-125": {
      "__comp": "__comp---theme-doc-item-178-a40",
      "__context": {
        "plugin": "plugin---docs-hello-665-3ca",
      },
      "content": "content---docs-foo-baz-8-ce-61e",
      "metadata": "metadata---docs-foo-baz-2-cf-fa7",
    },
  },
  "routesConfig": "import React from 'react';
import ComponentCreator from '@docusaurus/ComponentCreator';

export default [
  {
    path: '/docs:route',
    component: ComponentCreator('/docs:route', '001'),
    routes: [
      {
        path: '/docs/hello',
        component: ComponentCreator('/docs/hello', 'ab4'),
        exact: true,
        sidebar: "main"
      },
      {
        path: 'docs/foo/baz',
        component: ComponentCreator('docs/foo/baz', '125'),
        sidebar: "secondary",
        "key:a": "containing colon",
        "key'b": "containing quote",
        "key\\"c": "containing double quote",
        "key,d": "containing comma",
        字段: "containing unicode"
      }
    ]
  },
  {
    path: '*',
    component: ComponentCreator('*'),
  },
];
",
}
`;

exports[`loadRoutes loads route config with empty (but valid) path string 1`] = `
{
  "registry": {
    "__comp---hello-world-jse-0-f-b6c": "hello/world.js",
  },
  "routesChunkNames": {
    "-b2a": {
      "__comp": "__comp---hello-world-jse-0-f-b6c",
    },
  },
  "routesConfig": "import React from 'react';
import ComponentCreator from '@docusaurus/ComponentCreator';

export default [
  {
    path: '',
    component: ComponentCreator('', 'b2a')
  },
  {
    path: '*',
    component: ComponentCreator('*'),
  },
];
",
}
`;
